beater/cache.go

```diff
@@ -0,0 +1,89 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
+)
+
+const burstMultiplier = 5
+
+var limiterPool sync.Pool
+
+type rlCache struct {
+	cache *lru.Cache
+	limit int
+	mu    sync.Mutex //guards limiter in cache
+}
+
+func NewRlCache(size, rateLimit int) (*rlCache, error) {
+	if size <= 0 || rateLimit < 0 {
+		return nil, errors.New("Cache initialization went wrong.")
+	}
+
+	var onEvicted = func(key interface{}, value interface{}) {
+		limiterPool.Put(*value.(**rate.Limiter))
+	}
+
+	c, err := lru.NewWithEvict(size, onEvicted)
+	if err != nil {
+		return nil, err
+	}
+	rlc := rlCache{cache: c, limit: rateLimit}
+	return &rlc, nil
+}
+
+func (rlc *rlCache) getRateLimiter(key string) *rate.Limiter {
+	// fetch the rate limiter from the cache, if a cache is given
+	if rlc.cache == nil || rlc.limit == -1 {
+		return nil
+	}
+
+	getLimiter := func() (*rate.Limiter, bool) {
+		if l, ok := rlc.cache.Get(key); ok {
+			return *l.(**rate.Limiter), true
+		}
+		return nil, false
+	}
+
+	if limiter, ok := getLimiter(); ok {
+		return limiter
+	}
+
+	rlc.mu.Lock()
+	defer rlc.mu.Unlock()
+	if limiter, ok := getLimiter(); ok {
```

Comment:
I took a look at the LRU implementation. The LRU we use here is basically a wrapper around 	"github.com/hashicorp/golang-lru/simplelru" with locks added. In other words, there's a lock being taken for each call to `Get` and `Add`. In the last part of this method we're adding our own lock on top. I seems like we might as well as use `simplelru` and use our own lock?

I haven't thought this through completely, but instead of the get/lock/get/add/unlock pattern here, could you use  `ContainsOrAdd` by:
1. calling `Get`, if the IP is there the recentness gets updated and the limiter is returned
2. it was not there, call `ContainsOrAdd`, if it's there now, we'll return